src/Notmuch/Binding.chs

```diff
@@ -237,17 +245,69 @@ database_get_version :: Database a -> IO Int
 database_get_version db =
   fromIntegral <$> withDatabase db {#call unsafe database_get_version #}
 
+-- | Index a file with the default indexing options.
+-- (This binding does not yet provide a way to change
+-- the indexing options.)  Returns the indexed message.
+--
+-- If message has same message ID as another message in the
+-- database, the new filename will be added to the message
+-- and the existing message is returned.
+--
+-- Possible errors include:
+--
+-- * 'StatusFileError' when file does not exist or cannot be opened
+-- * 'StatusFileNotEmail' when file does not look like an email
+--
+indexFile
+  :: (AsNotmuchError e, MonadError e m, MonadIO m)
+  => Database RW -> FilePath -> m (Message 0 RW)
+indexFile db@(Database (DatabaseHandle dfp)) path =
+  liftIO (withDatabase db $ \db' ->
+    withCString path $ \path' ->
+      constructF'
+        (\r -> r == StatusSuccess || r == StatusDuplicateMessageId)
+        Identity
+        (fmap (Message [castForeignPtr dfp] . MessageHandle)
+          . (newForeignPtr notmuch_message_destroy <=< detachPtr))
+        ({#call unsafe database_index_file #} db' path' nullPtr)
+  )
+  >>= throwOr (pure . runIdentity)
+
+-- | Result of a 'removeFile' operation.
+data RemoveResult = MessageRemoved | MessagePersists
+  deriving (Eq, Show)
+
+-- | Remove a message filename.  If the message has no more
+-- filenames return 'MessageRemoved', otherwise 'MessagePersists'.
+--
+-- The underlying routine (as of notmuch v0.28) returns
+-- @NOTMUCH_STATUS_SUCCESS@ even when the given path does not
+-- exist, is not an internet message, or is not recorded in the
+-- database.  Therefore @removeFile@ also returns 'MessageRemoved'
+-- in this scenario.  This is particularly confusing when the
+-- @Message-Id@ of the given file is known, but the the file itself
+-- is unknown.
+--
+removeFile
+  :: (AsNotmuchError e, MonadError e m, MonadIO m)
+  => Database RW -> FilePath -> m RemoveResult
+removeFile db path =
+  liftIO (withDatabase db $ \db' ->
+    withCString path $ \path' ->
+      toStatus <$> {#call unsafe database_remove_message #} db' path'
+  )
+  >>= \e -> case e of
+    StatusSuccess -> pure MessageRemoved
+    StatusDuplicateMessageId -> pure MessagePersists
```

Comment:
I'm currently wondering what - as a user - you would do in this case. Ignore the error? I guess the database is in an inconsistent state, but what to do to fix it... Reason why I'm asking is: that's what I'd put in the error message as advice.

Reply:
It's not actually an error at all.  It just signals one of two outcomes: either there are still files associated with the message, in which case `MessagePersists`, otherwise there are no more files and the message was `MessageRemoved` from the index.

In our use scenario I think we can just ignore the result.  But I did *hs-notmuch* to signal this information to the caller in case it is important for the use case.